doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+ARM's SVE, RISC-V's RVV).
+
+As an example of the readability improvement, here is a typical line of C code
+from our current C universal intrinsics framework:
+
+.. code::
+
+   // The @name@ is the numpy-specific templating in .c.src files
+   npyv_@sfx@  a5 = npyv_load_@sfx@(src1 + npyv_nlanes_@sfx@ * 4);
+
+This will change (as implemented in PR `gh-21057`_) to:
+
+.. code:: C++
+
+   auto a5 = Load(src1 + nlanes * 4);
+
+If the above C++ code were to use Highway under the hood it would look quite
+similar, it uses similarly understandable names as ``Load`` for individual
+portable intrinsics.
+
+The ``@sfx`` in the C version above is the template variable for type
+identifiers, e.g.: ``#sfx = u8, s8, u16, s16, u32, s32, u64, s64, f32, f64#``.
+Explicit use of bitsize-encoded types like this won't work for sizeless SIMD
+instruction sets. With C++ this is easier to handle; PR `gh-21057`_ shows how
+and contains more complete examples of what the C++ code will look like.
+
+The scope of this NEP includes discussing all relevant aspects of adopting
+Google Highway to replace our current Universal Intrinsics framework, including
+but not limited to:
+
+- Maintainability, domain expertise availability, ease of onboarding new
+  contributor, and other social aspects,
+- Key technical differences and constraints that may impact NumPy's internal
+  design or performance,
+- Build system related aspects and impact on our plans for upstreaming CPU and
+  compiler feature detection and multi-target library building into Meson,
+- Release timing related aspects.
+
+Out of scope (at least for now) is revisiting other aspects of our current SIMD
+support strategy:
+
+- accuracy vs. performance trade-offs when adding SIMD support to a function
+- use of SVML and x86-simd-sort (and possibly its equivalents for aarch64)
+- pulling in individual bits or algorithms of Highway (as in `gh-24018`_) or
+  SLEEF (as discussed in that same PR)
+
+
+Usage and Impact
+----------------
+
+N/A - there will be no significant user-visible changes.
+
+
+Backward compatibility
+----------------------
+
+There will be no changes in user-facing Python or C APIs: all the methods to
+control compilation and runtime CPU feature selection should remain, although
+there may be some changes due to moving to Meson and C++ without regards to the
+Highway/Universal Intrinsics choice.
+
+On Windows, MSVC may have to be avoided, as a result of Highway's use of
+pragmas which are less well supported by MSVC. This means that we likely have
+to build our wheels with clang-cl or Mingw-w64. Both of those should work - we
+merged clang-cl support a while back (see `gh-20866`_), and SciPy builds with
+Mingw-w64. It may however impact other redistributors or end users who build
+from source on Windows.
+
+The potential impact on licensing because of Highway's use of the Apache 2.0
+license needs discussing before Highway can be accepted. Or Highway would have
```

Comment:
I'll try to be more explicit. Different concerns:

1. Apache 2.0 and GPLv2 are incompatible (a fading concern over the years, but still may have some impact)
2. We advertise NumPy as BSD-3 licensed. Including fully compatibly-licensed code like MIT and public domain code is fine. If we include Apache 2.0 code, we have to advertise all of NumPy not as BSD-3 but as mixed-licensed.
3. In addition to (2), if it's about mixing code within files, the actual license tracking becomes quite complicated. 

Eventually there may become a point where the benefits of including Apache 2.0 licensed code outweighs any of these downsides. Maybe use of Highway is that point. But it's still an important part of the decision, and has some consequences which are at least annoying to have to deal with.

@rkern may want to add some more color to the above.

Reply:
Thanks. About GPL incompatibility: I believe that is only an issue in one direction, if Apache2 code (i.e. Highway itself) were to depend on GPLed code, which would require the whole to be released under GPL. See https://www.apache.org/licenses/GPL-compatibility.html.

AFAIK there is no issue if someone else is GPL licensed and includes Numpy, with or without Highway. I also believe there is no need to declare Numpy as mixed-license because Apache2 does not impose additional limitations on users.

Reply:
No, there is still an issue with GPLv2. Projects using GPLv2 do not allow Apache 2.0 code into their codebase because of clauses in the Apache 2.0 license. Putting Apache 2.0 code into numpy means that GPLv2 codebases cannot/will not be able to use numpy.

> Despite our best efforts, the FSF has never considered the Apache License to be compatible with GPL version 2, citing the patent termination and indemnification provisions as restrictions not present in the older GPL license. The Apache Software Foundation believes that you should always try to obey the constraints expressed by the copyright holder when redistributing their work.

Reply:
hm, I see. That is unfortunate. Do we know of any GPLv2 users?

Reply:
[Certainly](https://pypi.org/search/?c=License+%3A%3A+OSI+Approved+%3A%3A+GNU+General+Public+License+v2+%28GPLv2%29&o=&q=numpy). It's not a perfect search, but clicking through, I do see a number of recently-maintained projects that specify GPLv2 without the "any later version" language that would allow users to upgrade to GPLv3. Whether that's an intentional omission or not is unknown in each case, but I do see some that depend on libraries that [some](https://pypi.org/project/glmnet2/) depend on external libraries [that also omit it](https://github.com/cran/glmnet) and thus don't have much choice in the matter.

Reply:
I agree with @rkern: We don't want to muddy the waters and provide friction for people to include NumPy in their projects.

Reply:
Got it. I will check internally about permission for dual-licensing.

Reply:
FYI I am waiting to hear from our open-source licensing folks.

Reply:
Thank you! It's much appreciated!

Reply:
:) Good news, the initial opinion is positive. Sounds like dual-licensing ought to be possible, and as a courtesy, I've filed https://github.com/google/highway/issues/1562 to inform folks and ask if there are any concerns.
We will follow up to confirm. FYI I will be out of office tomorrow.

Reply:
It's looking good and I believe we will be able to move forward with dual-licensing. Let's wait another week for contributors to register any concerns (unlikely).

Reply:
To clarify: the [current patch](https://github.com/google/highway/pull/1605) proposes BSD-3 as the other license alternative. The [meeting agenda](https://hackmd.io/76o-IxCjQX2mOXO_wwkcpg?both) mentions MIT. I think we could still change it at this point, but when notifying Highway contributors we mentioned BSD-3. Is that also OK for you?

Reply:
Either way works, BSD-3 and MIT are pretty much interchangeable.

Reply:
Thanks for confirming. Good news, the change is now approved, submitted, and shows up on Github. It's official :)